dynamic_array.c

```diff
@@ -0,0 +1,340 @@
+#include "dynamic_array.h"
+
+// NOTE: When we destroy items, we call the destructor for them 
+// iff the destructor is not NULL.
+
+typedef void (*destr)(void *);
```

Comment:
I suspect your editor is not terminating the last line of files when saving.  POSIX defines a line as a sequence of zero or more non-newline characters, terminated by a newline.  Many programs (including compilers) will misbehave if the last line of a file is not terminated by a newline (a so-called "incomplete line" in POSIX terms)